src/core/memory.cc

```diff
@@ -155,31 +156,13 @@ __thread volatile int critical_alloc_section = 0;
 #include <numaif.h>
 #endif
 
-namespace seastar {
-
-struct allocation_site {
-    mutable size_t count = 0; // number of live objects allocated at backtrace.
-    mutable size_t size = 0; // amount of bytes in live objects allocated at backtrace.
-    mutable const allocation_site* next = nullptr;
-    saved_backtrace backtrace;
-
-    bool operator==(const allocation_site& o) const {
-        return backtrace == o.backtrace;
-    }
-
-    bool operator!=(const allocation_site& o) const {
-        return !(*this == o);
-    }
-};
-
-}
 
 namespace std {
 
 template<>
-struct hash<seastar::allocation_site> {
-    size_t operator()(const seastar::allocation_site& bi) const {
-        return std::hash<seastar::saved_backtrace>()(bi.backtrace);
+struct hash<seastar::memory::allocation_site> {
```

Comment:
It is unusual to have the `std::hash` specialization defined in a different file than the class declaration (it means in this case that consumers won't get the hash impl). Is there something specific to justify it here?

Reply:
No not really. I originally wasn't sure whether the `allocation_site` struct would be a good fit as is.

I can't see anything wrong with it so we can move this over as well (and I can also add some comments).